### PR TITLE
[SDK] Add `env` & `env_from` in client tune

### PR DESCRIPTION
**Pull request by shipengcheng1230**
**What this PR does / why we need it**:

The `env` and `env_from` cannot be directly set in the client `tune` method. Recently I found myself need to set up proxy environment variables for cross-region data access. Doing so inside the existing python objective function seems less desirable than having them set through these two specifications. 

This PR enables this capability. By default, these two are set to None which aligns with [V1Container](https://github.com/kubernetes-client/python/blob/36cfbe68a509d9b9d33395b22b6fa94d7d46c30f/kubernetes/client/models/v1_container.py#L89). Therefore, behavior is not modified when not setting them.

~Also, there was a missing of `cmaes` package when testing optuna suggestion. Therefore I added it in the optuna requirements.txt.~ (Rebased the master change)

### CODE REVIEW

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py [hunk 1]
```diff
@@ -172,6 +175,8 @@ def tune(
                 objective function.
             base_image: Image to use when executing the objective function.
             namespace: Namespace for the Experiment.
+            env: Environment variable(s) to be attached to each trial container.
+            env_from: Source(s) of environment variables to be populated in each trial container.
```
Comment (tenzen-y) [inline sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:179]: Can we mention the types like `resources_per_trial`?

**Review by tenzen-y — COMMENTED** [review]
This idea looks good.

/assign @andreyvelich

File: cmd/suggestion/optuna/v1beta1/requirements.txt [hunk 2]
```diff
@@ -2,3 +2,4 @@ grpcio>=1.41.1
 protobuf>=3.19.5, <=3.20.3
 googleapis-common-protos==1.53.0
 optuna>=3.0.0
+cmaes>=0.10.0
```
[the remark this answers is not in the document]
Comment (shipengcheng1230) [reply cmd/suggestion/optuna/v1beta1/requirements.txt]: It seems the `cmaes` is listed as an optional depedency. And in the [GitHub CI](https://github.com/kubeflow/katib/actions/runs/6693916899/job/18186107558?pr=2235) it says
```
ModuleNotFoundError: No module named 'cmaes'
```

I think a change like from `optuna>=3.0.0` to `optuna[cmaes]>=3.0.0` is probably needed for the test to pass.

[the remark this answers is not in the document]
Comment (tenzen-y) [reply cmd/suggestion/optuna/v1beta1/requirements.txt]: Oh, I see. Thanks for the clarifying. Let's specify `cmaes>=0.10.0`.

**Review by tenzen-y — COMMENTED** [review]
Otherwise lgtm.

File: cmd/suggestion/optuna/v1beta1/requirements.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreyvelich) [reply cmd/suggestion/optuna/v1beta1/requirements.txt:5]: @shipengcheng1230 @tenzen-y Did optuna recently change their dependancies to make `cmaes` optional ?

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py [hunk 3]
```diff
@@ -140,6 +141,8 @@ def tune(
         parameters: Dict[str, Any],
         base_image: str = constants.BASE_IMAGE_TENSORFLOW,
         namespace: Optional[str] = None,
+        env: Union[Dict[str, str], List[V1EnvVar], None] = None,
+        env_from: Union[V1EnvFromSource, List[V1EnvFromSource], None] = None,
```
Comment (andreyvelich) [inline sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:145]: Should we keep the name associated with Trials ? E.g. user could mislead this as environment for Algorithm Service.
E.g. name it as `trials_env`.
Also, I think we can allow to set `Dict[str, str]`,  `List[V1EnvVar]`, or `List[V1EnvFromSource]` in this parameter.

WDYT @shipengcheng1230 @tenzen-y @johnugeorge ?

File: cmd/suggestion/optuna/v1beta1/requirements.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (shipengcheng1230) [reply cmd/suggestion/optuna/v1beta1/requirements.txt:5]: This commit moves it: https://github.com/optuna/optuna/commit/78159b516ebb694f5fa8a03072a2b54e9c53ef33

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py (hunk 3 above)
Comment (shipengcheng1230) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:145]: I like `env_per_trial` which is more descriptive and consistent with `resources_per_trial`!

Comment (andreyvelich) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:145]: Yeah, I think so @shipengcheng1230, we can accept multiple types for `env_per_trial` parameter.
What do you think about this name @tenzen-y @johnugeorge ?

Comment (shipengcheng1230) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py]: Here is the tentative implementation 12eea643d8ba49fd28e9be354941f80687b842e1

Comment (tenzen-y) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py]: `env_per_trial` looks great!

**Comment by tenzen-y** [comment]
Can you rebase this PR since we fixed the optuna dependency issue.

**Comment by shipengcheng1230** [comment]
Cool, done rebase and let me know what else is needed!

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py [hunk 4]
```diff
@@ -140,6 +141,7 @@ def tune(
         parameters: Dict[str, Any],
         base_image: str = constants.BASE_IMAGE_TENSORFLOW,
         namespace: Optional[str] = None,
+        env_per_trial: Union[Dict[str, str], List[Union[V1EnvVar, V1EnvFromSource]], None] = None,
```
Comment (andreyvelich) [inline sdk/python/v1beta1/kubeflow/katib/api/katib_client.py:144]: Should we make this type Optional and import libs from Kubernetes client, similar to [other Kubernetes modules](https://github.com/kubeflow/katib/blob/bab4e4d63a4a61340a749b54c809e2d71af78a3a/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py#L351) ?
```suggestion
        env_per_trial: Optional[Union[Dict[str, str], List[Union[client.V1EnvVar, client.V1EnvFromSource]]]] = None,
```

**Comment by andreyvelich** [comment]
Sorry for the late reply @shipengcheng1230, I left one small comment.

File: sdk/python/v1beta1/kubeflow/katib/api/katib_client.py (hunk 4 above)
Comment (shipengcheng1230) [reply sdk/python/v1beta1/kubeflow/katib/api/katib_client.py]: Thanks for the great suggestion, I wasn't aware that those `client.models` are also importable from the `client` level!

**Review by tenzen-y — COMMENTED** [review]
/approve

I'll leave lgtm on @andreyvelich.

**Comment by andreyvelich** [comment]
Thank you for the update @shipengcheng1230!
/lgtm

**Pull request merged**